aten/src/ATen/native/cuda/ReduceSumProdKernel.cu

```diff
@@ -35,6 +35,15 @@ struct prod_functor {
   }
 };
 
+template <typename scalar_t, typename acc_t=scalar_t, typename out_t=scalar_t>
+struct nanprod_functor {
+    void operator()(TensorIterator& iter) {
+        gpu_reduce_kernel<scalar_t, out_t>(iter, func_wrapper<out_t> ([]GPU_LAMBDA(acc_t a, acc_t b) -> acc_t {
+            return ((a != a) ? acc_t{1} : a) * (::isnan(b) ? acc_t{1} : b);
```

Comment:
Same question about cases. This is actually divergent from the CPU version of this function. Also, you probably don't want to mix the `a != a` pattern with the `::isnan` pattern.